src/Operation/RemoteMergeOperation.php

```diff
@@ -81,39 +80,83 @@ public function performMerge()
             throw new \RuntimeException('performMerge() was already called. Each operation is only usable once.');
         }
 
+        $this->performed = true;
+
         $this->gitHelper->stashBranchName();
-        $this->gitHelper->syncWithRemote($this->targetRemote, $this->targetBranch);
         $this->gitHelper->remoteUpdate($this->sourceRemote);
+        $this->gitHelper->remoteUpdate($this->targetRemote);
 
-        if ($this->switchBase) {
-            $this->gitHelper->syncWithRemote($this->targetRemote, $this->switchBase);
+        // To prevent pushing with an outdated base-branch (or pushing commits only existent
+        // in the local branch that has the same name as the remote-base) we checkout the
+        // "{targetRemote}/{targetBranch}" and create a temp-branch (temp-base).
+        //
+        // The temp-base equals the target-branch on the remote.
+        //
+        // Then we merge the temp-source branch into temp-base branch, and push explicitly
+        // to the remote base-branch with temp-base branch!
+        // In practice: 'git push {targetRemote} {temp-base}:{target-base}' is performed
+
+        $this->createBaseBranch();
+
+        $tempSourceBranch = $this->createSourceBranch();
+        $mergeHash = $this->gitHelper->mergeBranch($this->targetBase, $tempSourceBranch, $this->message);
+
+        $this->gitHelper->restoreStashedBranch();
+
+        return $mergeHash;
+    }
+
+    public function pushToRemote()
+    {
+        $target = trim($this->targetBase).':'.$this->targetBranch;
+
+        // Safety guard to prevent deleting a remote base branch!!
+        if (':' === $target[0]) {
+            throw new \RuntimeException(
+                sprintf('Push target "%s" does not include the local branch-name, please report this bug!', $target)
+            );
         }
 
+        $this->gitHelper->pushToRemote($this->targetRemote, $target);
+    }
+
+    private function createBaseBranch()
+    {
+        $targetBranch = null !== $this->switchBase ? $this->switchBase : $this->targetBranch;
+        $this->targetBase = $this->gitHelper->createTempBranch($this->targetRemote.'--'.$targetBranch);
```

Comment:
Then the process fails, It only happens when you abort the process before the command-terminate event is triggered.

We can add the microtime for extra randomness :+1: